File: python/socketd/socketd/transport/core/impl/ChannelBase.py
import asyncio
import threading

from abc import ABC
from typing import Optional

from socketd.transport.core import Entity
from socketd.transport.core.Channel import Channel
from socketd.transport.core.Costants import Constants
from socketd.transport.core.Frames import Frames
from socketd.transport.core.Config import Config
from socketd.transport.core.HandshakeDefault import HandshakeInternal
from socketd.transport.core.Message import Message
from socketd.utils.MapUtils import MapUtils


class ChannelBase(Channel, ABC):
    def __init__(self, config: Config):
        self.config = config

        self.handshake: HandshakeInternal = None
        self.__attachments:dict = {}

        self.__lock: threading.Lock = threading.Lock()
        self.__loop: Optional[asyncio.AbstractEventLoop] = None

    def __enter__(self):
        return self.__lock.acquire()

    def __exit__(self, exc_type, exc_val, exc_tb):
        return self.__lock.release()

    def get_config(self) -> Config:
        return self.config

    def is_valid(self) -> bool:
        pass

    def get_loop(self) -> asyncio.AbstractEventLoop:
        return self.__loop

    def set_loop(self, loop) -> None:
        self.__loop = loop

    def get_attachment(self, name):
        return self.__attachments.get(name, None)

    def put_attachment(self, name, val):
        if val is None:
            MapUtils.remove(self.__attachments, name)
        else:
            self.__attachments[name] = val

    def set_handshake(self, handshake):
        if handshake is not None:
            self.handshake = handshake

    def get_handshake(self):
        return self.handshake

    async def send_connect(self, uri, metaMap: dict[str, str]):
        await self.send(Frames.connect_frame(self.config.gen_id(), uri, metaMap), None)

    async def send_connack(self, connect_message):
        await self.send(Frames.connack_frame(self.get_handshake()), None)

    async def send_ping(self):
        await self.send(Frames.ping_frame(), None)

    async def send_pong(self):
        await self.send(Frames.pong_frame(), None)

    async def send_close(self, code: int):
        await self.send(Frames.close_frame(code), None)

    async def send_alarm(self, _from: Message, alarm: Entity):
        await self.send(Frames.alarm_frame(_from, alarm), None)

    async def send_pressure(self, _from: Message, pressure: Entity):
        await self.send(Frames.pressure_frame(_from, pressure), None)

    async def close(self, code: int):
        if code > Constants.CLOSE1000_PROTOCOL_CLOSE_STARTING:
            self.__attachments.clear()

File: javascript/socketd/src/socketd/transport/core/impl/ChannelBase.ts
import type {Config} from "../Config";
import type {HandshakeInternal} from "../Handshake";
import {Constants} from "../Constants";
import {type Frame} from "../Frame";
import type {Message} from "../Message";
import {SocketAddress} from "../SocketAddress";
import type {StreamInternal} from "../../stream/Stream";
import type {Session} from "../Session";
import {Channel} from "../Channel";
import {Frames} from "./Frames";
import {Entity} from "../Entity";

export abstract class  ChannelBase implements Channel {
    protected _config: Config;
    private _attachments: Map<string, any>;
    private _handshake: HandshakeInternal;

    constructor(config: Config) {
        this._config = config;
        this._attachments = new Map<string, object>();
    }

    abstract getLiveTime(): number;

    getAttachment<T>(name: string): T | null {
        return this._attachments.get(name);
    }

    putAttachment(name: string, val: object | null) {
        if (val == null) {
            this._attachments.delete(name);
        } else {
            this._attachments.set(name, val);
        }
    }

    abstract isValid(): boolean;

    abstract isClosing(): boolean;

    abstract closeCode(): number;

    close(code: number) {
        if (code > Constants.CLOSE1000_PROTOCOL_CLOSE_STARTING) {
            this._attachments.clear();
        }
    }


    getConfig(): Config {
        return this._config;
    }

    setHandshake(handshake: HandshakeInternal) {
        this._handshake = handshake;
    }

    getHandshake(): HandshakeInternal {
        return this._handshake;
    }

    sendConnect(url: string, metaMap: Map<string, string>) {
        this.send(Frames.connectFrame(this.getConfig().genId(), url, metaMap), null)
    }

    sendConnack() {
        this.send(Frames.connackFrame(this.getHandshake()), null);
    }

    sendPing() {
        this.send(Frames.pingFrame(), null);
    }

    sendPong() {
        this.send(Frames.pongFrame(), null);
    }

    sendClose(code:number) {
        this.send(Frames.closeFrame(code), null);
    }

    sendAlarm(from: Message, alarm: Entity) {
        this.send(Frames.alarmFrame(from, alarm), null);
    }

    abstract getRemoteAddress(): SocketAddress | null ;

    abstract getLocalAddress(): SocketAddress | null ;

    abstract send(frame: Frame, stream: StreamInternal<any> | null);

    abstract reconnect();

    abstract onError(error: any);

    abstract getSession(): Session ;
}